### ISSUE TITLE
[Question] How to create batch-independent GP with independent inputs

### ISSUE THREAD

**Issue by maciejkorzepa** [cornellius-gp/gpytorch#1183]
I'm trying to run the batch-independent GP example with a modification that the provided inputs are also batched, i.e. each GP can get different inputs (but the same number of them). However, providing an extra batch dimension for the input points leads to an error during prediction time.

The code to reproduce the issue:
```python
import math
import torch
import gpytorch

train_x = torch.linspace(0, 1, 100)

train_y = torch.stack([
    torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    torch.cos(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
    torch.cos(1 + train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2,
], -1)

train_x = train_x.repeat(3, 1).unsqueeze(-1) 

class BatchIndependentMultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, n_tasks):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean(batch_shape=torch.Size([n_tasks]))
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([n_tasks])),
            batch_shape=torch.Size([n_tasks])
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultitaskMultivariateNormal.from_batch_mvn(
            gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
        )


likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=3)
model = BatchIndependentMultitaskGPModel(train_x, train_y, likelihood, n_tasks=3)

model.train()
output = model(train_x)
model.eval()
output = model(train_x)
```
And the error trace:
```
/home/mjko/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/models/exact_gp.py:284: GPInputWarning: The input matches the stored training data. Did you forget to call model.train()?
  GPInputWarning,
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-f5bdb4e7ef08> in <module>
     36 output = model(train_x)
     37 model.eval()
---> 38 output = model(train_x)

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/models/exact_gp.py in __call__(self, *args, **kwargs)
    326             # Make the prediction
    327             with settings._use_eval_tolerance():
--> 328                 predictive_mean, predictive_covar = self.prediction_strategy.exact_prediction(full_mean, full_covar)
    329 
    330             # Reshape predictive mean to match the appropriate event shape

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/models/exact_prediction_strategies.py in exact_prediction(self, joint_mean, joint_covar)
    296             test_train_covar = test_covar[..., : self.num_train]
    297         else:
--> 298             test_test_covar = joint_covar[..., self.num_train :, self.num_train :]
    299             test_train_covar = joint_covar[..., self.num_train :, : self.num_train]
    300 

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/lazy_tensor.py in __getitem__(self, index)
   1713             res = self._get_indices(row_index, col_index, *batch_indices)
   1714         else:
-> 1715             res = self._getitem(row_index, col_index, *batch_indices)
   1716 
   1717         # If we selected a single row and/or column (or did tensor indexing), we'll be retuning a tensor

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/block_lazy_tensor.py in _getitem(self, row_index, col_index, *batch_indices)
     96 
     97         # Now construct a kernel with those indices
---> 98         return self.__class__(new_base_lazy_tensor, block_dim=-3)
     99 
    100     def _matmul(self, rhs):

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/block_lazy_tensor.py in __init__(self, base_lazy_tensor, block_dim)
     30 
     31     def __init__(self, base_lazy_tensor, block_dim=-3):
---> 32         if base_lazy_tensor.dim() < 3:
     33             raise RuntimeError(
     34                 "base_lazy_tensor must be a batch matrix (i.e. at least 3 dimensions - got "

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/lazy_tensor.py in dim(self)
    838         Alias of :meth:`~gpytorch.lazy.LazyTensor.ndimension`
    839         """
--> 840         return self.ndimension()
    841 
    842     @property

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/lazy_tensor.py in ndimension(self)
   1146         Returns the number of dimensions
   1147         """
-> 1148         return len(self.size())
   1149 
   1150     def numel(self):

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/lazy_tensor.py in size(self, val)
   1429         Returns the size of the resulting Tensor that the lazy tensor represents
   1430         """
-> 1431         size = self._size()
   1432         if val is not None:
   1433             return size[val]

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/utils/memoize.py in g(self, *args, **kwargs)
     32         cache_name = name if name is not None else method
     33         if not is_in_cache(self, cache_name):
---> 34             add_to_cache(self, cache_name, method(self, *args, **kwargs))
     35         return get_from_cache(self, cache_name)
     36 

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/lazy/lazy_evaluated_kernel_tensor.py in _size(self)
    213                     self.kernel.batch_shape,
    214                     error_msg=(
--> 215                         f"x1 and x2 were not broadcastable with kernel of batch_shape {self.kernel.batch_shape}. "
    216                         f"Got x1.shape = {x1.shape} and x2.shape = {x2.shape}"
    217                     ),

~/.conda/envs/env_torch/lib/python3.7/site-packages/gpytorch/utils/broadcasting.py in _mul_broadcast_shape(error_msg, *shapes)
     20                     raise RuntimeError("Shapes are not broadcastable for mul operation")
     21                 else:
---> 22                     raise RuntimeError(error_msg)
     23             final_size.append(non_singleton_sizes[0])
     24         # In this case - all dimensions are singleton sizes

RuntimeError: x1 and x2 were not broadcastable with kernel of batch_shape torch.Size([3]). Got x1.shape = torch.Size([0, 200, 1]) and x2.shape = torch.Size([0, 200, 1])
```

Interestingly, if I run with 2 tasks, I'm getting a different error (`Shapes are not broadcastable for mul operation`).

**Comment by jacobrgardner** [comment]
This is a bug in the still pretty new `from_batch_mvn` interface. For now, you can accomplish this using the original batch GP interface here: https://github.com/cornellius-gp/gpytorch/blob/master/examples/08_Advanced_Usage/Simple_Batch_Mode_GP_Regression.ipynb

**Comment by gpleiss** [comment]
@jacobrgardner I'm not sure this is a bug. For a MultitaskGP, we shouldn't want a batch of inputs (each batch item corresponding to a separate task).

 @maciejkorzepa - if I'm understanding your use case - you want a batch of GPs, but not necessarily a multitask model, right? If that is the case, then I think the example that @jacobrgardner  points to is the correct example to follow; not the BatchIndependentMultitaskGP example.

**Comment by maciejkorzepa** [comment]
@jacobrgardner @gpleiss Thanks for help! Yes, I don't need a multitask model here, so the example with Batch mode GPs you linked to seem to be exactly what I need.

Let me ask one more question related to batch GPs. I'm working on a project where I have M datasets (M can be as large as a few thousands) and each dataset consists of a couple of thousands of data points. For each dataset I want to train an exact GP (with fixed hyperparameters, so I don't need to optimize marginal likeilhood). Naturally, it's way too much data to fit on GPU and train all GPs (5000 datasets with 2500 data points each already sum up to ~23GB of training kernels, even more is needed for inference). I need to fit the models only once, then I use them only for predictions. I would like to be able to efficiently evaluate them in parallel (sometimes all and sometimes only a subset of them). What would be a good way to implement it efficiently? Using LOVE could reduce memory and computation time if I'm correct? However, even so, I won't be able to train them at once. Maybe training them in subsets and then combining into one would be a feasible option?

**Comment by jacobrgardner** [comment]
@maciejkorzepa Try the vanilla batch GP with a KeOps kernel, e.g. https://github.com/cornellius-gp/gpytorch/blob/master/examples/02_Scalable_Exact_GPs/KeOps_GP_Regression.ipynb or our built in kernel partitioning / checkpointing.

You shouldn't have much trouble training all of them in a single model using MVM partitioning. If you do, honestly it probably won't be that much slower to just train like 2 models, each with 2500 of the datasets. You're starting to work with large enough tensors there that it's gonna saturate the GPU anyways. If you have multiple GPUs, you could do a single model split across multiple GPUs: https://github.com/cornellius-gp/gpytorch/blob/master/examples/02_Scalable_Exact_GPs/Simple_MultiGPU_GP_Regression.ipynb
